src/platform/test/rom-test-main.c

```diff
@@ -0,0 +1,219 @@
+
+#include <mgba/core/cheats.h>
+#include <mgba/core/config.h>
+#include <mgba/core/core.h>
+#include <mgba/core/serialize.h>
+#include <mgba/internal/gba/gba.h>
+
+#include <mgba/feature/commandline.h>
+#include <mgba-util/vfs.h>
+
+#include <signal.h>
+
+#define ROM_TEST_OPTIONS "S:R:"
+#define ROM_TEST_USAGE \
+	"\nAdditional options:\n" \
+	"  -S SWI           Run until specified SWI call before exiting\n" \
+	"  -R REGISTER      General purpose register to return as exit code\n" \
+
+struct RomTestOpts {
+	int exitSwiImmediate;
+	unsigned int returnCodeRegister;
+};
+
+static void _romTestShutdown(int signal);
+static bool _parseRomTestOpts(struct mSubParser* parser, int option, const char* arg);
+static bool _parseSwi(const char* regStr, int* oSwi);
+static bool _parseNamedRegister(const char* regStr, unsigned int* oRegister);
+
+static bool _dispatchExiting = false;
+static int _exitCode = 0;
+
+#ifdef M_CORE_GBA
+static void _romTestSwi3Callback(struct mCore* core);
+
+static void _romTestSwi16(struct ARMCore* cpu, int immediate);
+static void _romTestSwi32(struct ARMCore* cpu, int immediate);
+
+static int _exitSwiImmediate;
+static unsigned int _returnCodeRegister;
+
+void (*_armSwi16)(struct ARMCore* cpu, int immediate);
+void (*_armSwi32)(struct ARMCore* cpu, int immediate);
+#endif
+
+int main(int argc, char * argv[]) {
+	signal(SIGINT, _romTestShutdown);
+
+	struct RomTestOpts romTestOpts = { 3, 0 };
+	struct mSubParser subparser = {
+		.usage = ROM_TEST_USAGE,
+		.parse = _parseRomTestOpts,
+		.extraOptions = ROM_TEST_OPTIONS,
+		.opts = &romTestOpts
+	};
+
+	struct mArguments args;
+	bool parsed = parseArguments(&args, argc, argv, &subparser);
+	if (!args.fname) {
+		parsed = false;
+	}
+	if (!parsed || args.showHelp) {
+		usage(argv[0], ROM_TEST_USAGE);
+		return !parsed;
+	}
+	if (args.showVersion) {
+		version(argv[0]);
+		return 0;
+	}
+	struct mCore* core = mCoreFind(args.fname);
+	if (!core) {
+		return 1;
+	}
+	core->init(core);
+	mCoreInitConfig(core, "romTest");
+	applyArguments(&args, NULL, &core->config);
+
+	mCoreConfigSetDefaultValue(&core->config, "idleOptimization", "remove");
+
+#ifdef M_CORE_GBA
+	if (core->platform(core) == mPLATFORM_GBA) {
+		((struct GBA*) core->board)->hardCrash = false;
+
+		_exitSwiImmediate = romTestOpts.exitSwiImmediate;
+		_returnCodeRegister = romTestOpts.returnCodeRegister;
+
+		if (_exitSwiImmediate == 3) {
+			// Hook into SWI 3 (shutdown)
+			struct mCoreCallbacks callbacks = {0};
+			callbacks.context = core;
+			callbacks.shutdown = (void(*)(void*)) _romTestSwi3Callback;
+			core->addCoreCallbacks(core, &callbacks);
+		} else {
+			// Custom SWI hooks
+			_armSwi16 = ((struct GBA*) core->board)->cpu->irqh.swi16;
+			((struct GBA*) core->board)->cpu->irqh.swi16 = _romTestSwi16;
+			_armSwi32 = ((struct GBA*) core->board)->cpu->irqh.swi32;
+			((struct GBA*) core->board)->cpu->irqh.swi32 = _romTestSwi32;
+		}
+	}
+#endif
+
```

Comment:
It might be good to add an "unsupported" for other here, otherwise it'll just run forever (or add another way to exit, such as a maximum frame count). Not a huge deal for the initial PR though. I can add this after it's merged.